libsolidity/ast/Types.h

```diff
@@ -1044,7 +1073,7 @@ class EnumType: public Type
  * Type that can hold a finite sequence of values of different types.
  * In some cases, the components are empty pointers (when used as placeholders).
  */
-class TupleType: public Type
+class TupleType: public CompositeType
```

Comment:
Is there a way to get a storage collision warning with tuples?

Reply:
I tried just now but it looks like the members of the tuple are are checked only individually. So for example a tuple like this is not considered too large for storage:

```solidity
contract C {
    uint[2**256-1] z;

    function f() public {
        (uint[2**256-1] storage x, uint[2**256-1] storage y) = (z, z);
    }
}
```

Same happens for tuples of memory types. And interestingly, for structs too:

```solidity
pragma experimental ABIEncoderV2;

struct S {
    uint[2**27-1] s1;
    uint[2**27-1] s2;
}

contract C {
    uint[2**256-1] z;

    function g() public {
        abi.decode("abc", (uint[2**27-1], uint[2**27-1]));
    }

    function h() public {
        abi.decode("abc", (S));
    }
}
```

There's a complication though - tuples can have mixed memory/storage content:
```solidity
contract C {
    uint[2**256-1] z;

    function f() public {
        uint[2**27-1] memory w;
        (uint[2**256-1] storage x, uint[2**27-1] memory y) = (z, w);
    }
}
```

Reply:
> tuples can have mixed memory/storage content

Oh, indeed! Perhaps we should ignore tuples. I think they are not real types, but rather some sophisticated syntax for grouped assignments.

Reply:
Well, @ekpyron would probably have something to say here :) They might become a real type one day but right now they're just some weird special case.

I mentioned them mostly for completeness and because on the surface they do seem to fit the definition of a composite type but maybe we should really leave them alone :)

Reply:
Would I :-)? "Tuples are weird" maybe :-)...
But the checks here will only be applied to types that declare variables, won't they? And you can't declare a variable of tuple type, so the checks shouldn't be affected no matter how we treat tuples here...
Conceptually they would probably count as "CompositeTypes"... but yeah, they're not really full fletched types...

So I'd be fine with both treating tuples as ``CompositeTypes`` and ignoring them - with the tendency that treating them as ``CompositeTypes`` would be more consistent, especially if it indeed doesn't affect the checks negatively.

Reply:
I agree with the consistency argument. However I feel a bit uneasy to add unused new code, with a hope that it will work when the day comes. Would [a comment and an assert](https://github.com/ethereum/solidity/pull/9818/files#diff-c903efe2e4a71d75dc2eed4252cb24d8R1102-R1107) be an option?

Reply:
I think it would be a good compromise for the time being.